src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java

```diff
@@ -145,15 +147,24 @@ void updateSimulationStats() {
 
     /** Write final simulation statistical data to db. */
     void writeSimulationStats() {
-        for (Beacon beacon : beacons) {
-            beaconsObservedSum.putIfAbsent(String.valueOf(beacon.getId()), 0D);
-        }
-        Map<String, Double> beaconsObservedPercent = beaconsObservedSum.entrySet().stream()
-                .collect(toImmutableMap(e -> e.getKey(), e -> e.getValue() / (currentRound - 1)));
+        Map<String, List<Integer>> observedIntervals = beaconsObservedIntervals.entrySet().stream()
+                .collect(toImmutableMap(e -> String.valueOf(e.getKey().getId()),
+                e -> e.getValue().stream().filter(i -> i.observed()).map(ObservedInterval::duration).collect(toImmutableList())));
+
+        Map<String, List<Integer>> unobservedIntervals = beaconsObservedIntervals.entrySet().stream()
+                .collect(toImmutableMap(e -> String.valueOf(e.getKey().getId()),
+                e -> e.getValue().stream().filter(i -> !i.observed()).map(ObservedInterval::duration).collect(toImmutableList())));
+
+        Map<String, Double> beaconsObservedPercent = observedIntervals.entrySet().stream() // delete after deletion from StatsState
+                .collect(toImmutableMap(e -> e.getKey(), // delete after deletion from StatsState
+                e -> e.getValue().stream().mapToDouble(Double::valueOf).sum() / (currentRound - 1))); // delete after deletion from StatsState
```

Comment:
Consider adding a helper function for this partial sum (for better readability).

Reply:
(Alternatively: skip the `mapToDouble()`, summing the integer intervals as-is, and cast `(currentRound - 1)` to `(double)(currentRound - 1)` - Java will promote the integer sum to double, and the code is slightly shorter)